### PR TITLE
feat: refine generator types for co

### PR DESCRIPTION
**Pull request by upupming**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [ ] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [ ] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [ ] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [ ] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [ ] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [ ] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: example from https://github.com/tj/co/blob/master/Readme.md
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

If removing a declaration:
- [ ] If a package was never on Definitely Typed, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by doniyor2109** [comment]
@upupming Is it going to be breaking change for users who use Typescript 3.1?

**Comment by upupming** [comment]
> @upupming Is it going to be breaking change for users who use Typescript 3.1?

@doniyor2109 Yes, I have used the new generator feature in Typescript 3.6 (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-6.html) as commented in the source code. The main change is to refine the `ExtractType` util and the `co` supports `Iterator<any, any, any>` instead of only `Iterator<any>` as arguments. 

Using the `Generator` type, we can know exactly the return type and yielded the type of the generator. See the example below:

```ts

type ExtractType<I> = I extends { [Symbol.iterator]: () => Iterator<any, infer TReturn, any> } ? TReturn :
  I extends (...args: any[]) => any ? ReturnType<I> : I

// yield number, return string, next boolean
type A = ExtractType<Generator<number, string, boolean>>
// yield number, no return, next boolean
type B = ExtractType<Generator<number, undefined, boolean>>
// no yield number, return string, next boolean
type C = ExtractType<Generator<undefined, string, boolean>>

export type extractTypeCases = [
  // All we got is just TReturn type
  Expect<Equal<A, string>>,
  Expect<Equal<B, undefined>>,
  Expect<Equal<C, string>>,
]
```

**Comment by doniyor2109** [comment]
@upupming Then we should ensure so that @types/co will be bumped to minor version or major version. I am not sure how to tell @typescript-bot to bump it with minor version or major version. 

Can we do that as well?

**Comment by upupming** [comment]
I have found this information: https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-use-features-from-very-new-typescript-versions

it says:

> Then you will have to add a comment to the last line of your definition header (after // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped): // Minimum TypeScript Version: X.Y. This will set the lowest minimum supported version.

**Review by doniyor2109 — APPROVED** [review]

File: types/co/index.d.ts [hunk 1]
```diff
@@ -4,18 +4,25 @@
 // Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
 // TypeScript Version: 3.1
```
[the remark this answers is not in the document]
Comment (upupming) [reply types/co/index.d.ts]: Thanks for the review, I have updated this file before your review, is it okay to merge now? It seems your approval is out of date.

**Comment by upupming** [comment]
@doniyor2109 I have also found [this](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library):

> The DefinitelyTyped-tools (the tool that publishes @types packages to npm) will set the declaration package's version by using the major.minor version number listed in the first line of its index.d.ts file.
> Note that the first-line comment in the index.d.ts file should only contain the major.minor version (e.g. 10.12) and should not contain a patch version (e.g. 10.12.4).

It seems type definitions must keep the same `major.minor` version with the corresponding JS package. So we cannot change the current `4.6` (the latest co version) to `4.7`. 

image

**Comment by doniyor2109** [comment]
> @doniyor2109 I have also found [this](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library):
> 
> > The DefinitelyTyped-tools (the tool that publishes @types packages to npm) will set the declaration package's version by using the major.minor version number listed in the first line of its index.d.ts file.
> > Note that the first-line comment in the index.d.ts file should only contain the major.minor version (e.g. 10.12) and should not contain a patch version (e.g. 10.12.4).
> 
> It seems type definitions must keep the same `major.minor` version with the corresponding JS package. So we cannot change the current `4.6` (the latest co version) to `4.7`.
> 
> image

@upupming  Yea, that is unfortunate. If there is no way to do that, let's leave it as it is then. At least we specified minimum typescript version. 

If someone wants to update to new version of @types/co then they will also update TS version. I think migrating from 3.1 to 3.6 version should not be problematic if it is not big project). If upgrading TS version is not possible, they can use old version for now.

**Review by doniyor2109 — APPROVED** [review]

**Comment by upupming** [comment]
Ready to merge

**Pull request merged**